### PR TITLE
feat(opensearchservice): SAML authorization properties for Domain construct

### LINKED ISSUE #26600: aws_opensearchservice: High Level Constructs For OpenSearch SamlOptions Feature
### Describe the feature

SAMLOptions is an existing Feature for OpenSearch Domain. [Feature Documentation Link](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/saml.html). According to the CDK docs currently there is no high level CDK construct for this feature.[CDK Docs link for the SamlOptions](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_opensearchservice.CfnDomain.SAMLOptionsProperty.html)

### Use Case

The general recommendation is to use high level constructs and due to lack of High level construct for this feature, the CDK template needs to be migrated to use CFN constructs if there is a need to Create an OpenSearchDomain with SAMLOptionsProperty. This serves as a hinderance to the adoption of the SAMLOptionsProperty feature for AWS OpenSearch Domains.

### Proposed Solution

Add the SAMLOptionsProperty to the AdvancedSecurityOptions interface object [CDK Doc for AdvancedSecurityOptions Attribute](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_opensearchservice.AdvancedSecurityOptions.html) .
A SamlOptions enabeld OpenSearchDomain created using CDK high level constructs should look like
```
const domain = new Domain(this, 'Domain', {
  version: EngineVersion.OPENSEARCH_1_0,
  enforceHttps: true,
  nodeToNodeEncryption: true,
  encryptionAtRest: {
    enabled: true,
  },
  fineGrainedAccessControl: {
    masterUserName: 'master-user',
    samlOptions: {
      enabled: true,
      idp: {
             entityId: 'entityId',
             metadataContent: 'metadataContent',
      },
     masterBackendRole: 'masterBackendRole',
     masterUserName: 'masterUserName',
     rolesKey: 'rolesKey',
     sessionTimeoutMinutes: 123,
     subjectKey: 'subjectKey',
   }
  },
  logging: {
    auditLogEnabled: true,
    slowSearchLogEnabled: true,
    appLogEnabled: true,
    slowIndexLogEnabled: true,
  },
});
```

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.84.0

### Environment details (OS name and version, etc.)

macOS Ventura 13.4

### PR DESCRIPTION
**Pull request by lpizzinidev**
Allows to specify [SAML authentication](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/saml.html) for OpenSearch domains via high-level construct properties.

Example:
```
const domain = new Domain(this, 'Domain', {
  version: EngineVersion.OPENSEARCH_1_0,
  enforceHttps: true,
  nodeToNodeEncryption: true,
  encryptionAtRest: {
    enabled: true,
  },
  fineGrainedAccessControl: {
    masterUserName: 'master-user',
    samlAuthenticationEnabled: true,
    samlAuthenticationOptions: {
      idpEntityId: 'entity-id',
      idpMetadataContent: 'metadata-content-with-quotes-escaped',
    },
  },
});
```

Closes #26600.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts [hunk 1]
```diff
@@ -304,6 +353,20 @@ export interface AdvancedSecurityOptions {
...
+   * @see https://docs.aws.amazon.com/opensearch-service/latest/developerguide/saml.html
+   *
+   * @default - SAML authentication is disabled. Enabled if `samlAuthenticationOptions` is set.
+   */
+  readonly samlAuthenticationEnabled?: boolean;
+
+  /**
+   * Container for information about the SAML configuration for OpenSearch Dashboards.
+   */
+  readonly samlAuthenticationOptions?: SAMLOptionsProperty;
```
Comment (colifran) [inline packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts:369]: Could you add a default stating "no SAML authentication options"? - `@default - no SAML authentication options`

I also think we should add a note here saying that if SAML authentication options are set, then `samlAuthenticationEnabled` will be enabled

File: packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts [hunk 2]
```diff
@@ -1685,6 +1757,18 @@ export class Domain extends DomainBase implements IDomain, ec2.IConnectable {
             masterUserName: masterUserName,
             masterUserPassword: this.masterUserPassword?.unsafeUnwrap(), // Safe usage
           },
+          samlOptions: samlAuthenticationEnabled ? {
+            enabled: true,
+            idp: {
```
Comment (colifran) [inline packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts:1762]: How about something like this so we can remove the `!` operator:
```ts
idp: props.fineGrainedAccessControl && props.fineGrainedAccessControl.samlAuthenticationOptions ? {
  entityId: props.fineGrainedAccessControl.samlAuthenticationOptions.idpEntityId,
  metadataContent: props.fineGrainedAccessControl.samlAuthenticationOptions.idpMetadataContent,
} : undefined,
```
What do you think?

File: packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts [hunk 3]
```diff
@@ -276,6 +276,55 @@ export enum TLSSecurityPolicy {
...
+   * The metadata of the SAML application, in XML format.
+   */
+  readonly idpMetadataContent: string;
+
+  /**
+   * The SAML master user name, which is stored in the domain's internal user database.
+   * This SAML user receives full permission in OpenSearch Dashboards/Kibana.
+   * Creating a new master username does not delete any existing master usernames.
+   */
+  readonly masterUserName?: string;
```
Comment (colifran) [inline packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts:298]: Could you add a `default flag`?

File: packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts [hunk 4]
```diff
@@ -276,6 +276,55 @@ export enum TLSSecurityPolicy {
...
+
+  /**
+   * The backend role that the SAML master user is mapped to.
+   * Any users with this backend role receives full permission in OpenSearch Dashboards/Kibana.
+   */
+  readonly masterBackendRole?: string;
```
Comment (colifran) [inline packages/aws-cdk-lib/aws-opensearchservice/lib/domain.ts:304]: Could you add a `default` flag for this one too?

**Review by colifran — DISMISSED** [review]
@lpizzinidev This looks great! I left a few minor comments.

**Review by colifran — APPROVED** [review]
@lpizzinidev Looks great! Thanks for your work on this.

**Review by colifran — APPROVED** [review]

**Comment by rehanvdm** [comment]
@lpizzinidev @colifran I don't see the docs having updated but I see the README.md was updated in this PR. Do you know when/how the docs will update? 

image

**Pull request merged**